### PR TITLE
[BEAM-214] Add ParquetIO

### PR DESCRIPTION
**Pull request by jbonofre**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by jbonofre** [comment]
This is a first simple version of the `ParquetIO` dealing with Avro `GenericRecord`.
I have the following TODO list on this one:
* the `Write` (test) has to be fixed with the direct runner.
* improve (maybe providing a custom `ParquetReader`) the `Read` to use a `BlockBasedSource` to be able to split the read on different sources (using offset, ...).

**Comment by dhalperi** [comment]
Did you want a reviewer for this, or is this just to exhibit the code? If you did want a reviewer, R: me?

**Comment by jbonofre** [comment]
Hey Dan. The PR is not fully ready, but you can already take a look in your have some spare time ;) Thanks !

**Comment by davorbonaci** [comment]
Hi guys, just checking, any updates perhaps?

**Comment by jbonofre** [comment]
Resuming my work on all PRs

**Comment by Igosuki** [comment]
Hey dude, I'm already using this code for production, however I added some more to allow parameterizing. 
- Using specific datum encoding is necessary to support 1.8.1's logical type conversions.
- Parameterizing helps with that but also by not having to cast SpecificRecords in another SerDe step in your beam since the ParquetReader/Writer already handles that.

I also think you should at least document how to pass hadoop configuration properties since that's used by parquet and doesn't go through the usual beam options with these code.
I can help contribute some code if you want.

**Comment by jbonofre** [comment]
Hey !
Agree with your comments/changes ! Let me improve the PR !
You can also submit PRs on my branch !
I'm resuming my work on this PR (and also others ;)).
Thanks !

**Comment by DrIgor** [comment]
@Igosuki could you please share your improved version?

Thank you!

**Comment by LiQusha** [comment]
Any updates? IMHO, without parquet, batch mode (and/or spark runner) is mostly impractical. However, with parquet, Cloud Storage (as data lake) and DataFlow( with the new Apache Beam SQL DSL) could be a great combo.

**Comment by jbonofre** [comment]
I plan to resume my work on this PR later this week.

**Comment by nehalecky** [comment]
Also very interested in this PR, look forward to more updates!

**Comment by jbonofre** [comment]
I'm on the top of my TODO for the following days (I just updated `AmqpIO`). Stay tuned !

**Comment by davorbonaci** [comment]
Any updates perhaps?

**Comment by jbonofre** [comment]
Yup. Starting the rebasing and update.

**Comment by nehalecky** [comment]
@jbonofre, thanks for the update!

**Comment by nielsbasjes** [comment]
I did some analysis on what goes wrong writing to a Parquet file.
First I added the option present in the Parquet write to overwrite the output.
That yielded a new insight into what goes wrong.
It turns out that the current implementation uses the same filename for each parallel instance writing the data. That is why the test fails with file already exists.

With the overwrite option you therfor get a different error that effectively says "bad file" which is simply caused by the threads obliterating the data written by the others.

My current assessment is that the "numShards and "shardNameTemplate" construct (as seen in TextIO) needs to be added here as well to make it work.

**Comment by nielsbasjes** [comment]
I did a local attempt to use FileBasedSink to make this work because that already has the notion of numShards. 
As far as I can tell this won't work for Parquet because I think that this assumes a header, set of records and a footer. The desired number of shards is later on realized by rewriting to the desired number of files after everything has been completed.
If I am right here then formats like Parquet and Orc (i.e. blocks of mangled records with advanced stats in headers and footers) will be very hard to created in a controlled way in Beam.

**Comment by jbonofre** [comment]
Rebased and updated to Beam 2.2.0-SNAPSHOT.

**Comment by jbonofre** [comment]
@nielsbasjes I tried to use block and file based sink/source first in `AvroIO`. Unfortunately, the Parquet API and Beam don't match so well ;), so I stayed at a more "higher" level.

**Comment by jbonofre** [comment]
retest this please

**Comment by jbonofre** [comment]
R: @jkff 

Eugene, can we chat a bit about this PR ? I would like to move forward on this one and I would love to have your input and comments. Thanks !

**Comment by jkff** [comment]
Sure, what would you like to discuss? I'd like to understand more about the use cases of Parquet in the wild to decide whether the proposed API in this PR is the right API - is there any Parquet expert you can add also as a reviewer? Eg I suppose people might want to choose to read only a subset of columns, or use filter pushdown. Write might need to be configurable in some way too.

Other notes to modernize this PR:

- The read transform seems to be non-parallel and reads a single file. Is this intentional?
- I would recommend to start right away with a readFiles() version, like in TextIO, so that this can be freely combined with FileIO.match() and FileIO.read(). Is it possible to read chunks of a single Parquet file in parallel? If yes, then might want to use ReadAllFiles and have a BlockBasedSource, but might also just start with a ReadFn (but on top of ReadableFile).
- Likewise I would recommend to implement the write side just as a sink on top of s.apache.org/fileio-write when it is ready (which should be soon enough so I think it makes sense to wait).

**Comment by Igosuki** [comment]
Parquet is used to serialize and compress proto, thrift and avro files into
queryable column oriented files.  The hard part is that it uses a very api
private internal reader to do a break down and parallelize reading. So the
code to split the input source like with other beam io would have to be
custom

**Comment by Igosuki** [comment]
A complete parquet implementation would allow pushing down the FilterApi and splitting the workload using the same kind of method that's used in io's like hadoop-file-system not the internal reader as seen here https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java#L141.
I think Beam should read the footers itself to determine how to parallelize the workload.

**Comment by jbonofre** [comment]
I've started to implement a `ReadFiles` wrapping `ReadAllViaFileBasedSource`. Once done, I will PoC to implement sharding based on the footer.

**Comment by jkff** [comment]
Is this PR in active development? Would it make sense to close it for now and reopen when there's something ready to be reviewed?

**Comment by jbonofre** [comment]
I'm working on readFiles important now. This code is already used by some users. If you don't mind, I prefer to keep this PR open as I think I'm pretty close of the readFiles impl.

**Comment by jbonofre** [comment]
Rebased and added the gradle build. I'm working on adding the `readAll` pattern support.

**Comment by jkff** [comment]
Alright, let me know when it's ready for review.

For now I have a few comments:

- The interface seems *very* similar to AvroIO. That means perhaps it should be pretty much identical (e.g. allow reading using Avro generated classes, allow not specifying the schema etc.) - WDYT? Or perhaps replicating the API service of AvroIO would be too much work - but then what would be the minimum useful subset?

- Splitting within a file appears to be lacking, which could be a serious performance issue. It would be good to have at least basic splitting into reasonably-sized chunks, even if not dynamic splitting (which would require BoundedSource and going via ReadViaFileBasedSource).

File: sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+ *
+ * <p>For example:
+ *
+ * <pre>{@code
+ *  pipeline
+ *    .apply(...) // PCollection<GenericRecord>
+ *    .apply(ParquetIO.write().withPath("/foo/bar").withSchema(schema));
+ * }</pre>
+ */
+public class ParquetIO {
```
Comment (jkff) [inline sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: Please mark this experimental and comment that the API will change.

File: sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    return new AutoValue_ParquetIO_Write.Builder().build();
+  }
+
+  /**
+   * A {@link PTransform} to read from Parquet file.
+   */
+  @AutoValue
+  public abstract static class Read extends PTransform<PBegin, PCollection<GenericRecord>> {
+
+    @Nullable abstract String path();
```
Comment (jkff) [inline sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: We should support expanding the filepattern. Maybe better then to implement this as a `PTransform<PCollection<ReadableFile>, PCollection<GenericRecord>>` much like `AvroIO.readFiles()` (I think that's what it's called).

File: sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+    }
+
+  }
+
+  /**
+   * {@link PTransform} writing a {@link PCollection} of {@link GenericRecord} into a Parquet file.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<GenericRecord>, PDone> {
```
Comment (jkff) [inline sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: This should be a `FileIO.Sink` to integrate with `FileIO.write()`.

**Comment by jbonofre** [comment]
I will resume my work on it today after RC2 on vote. Thanks !

File: sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java (hunk 1 above)
Comment (jbonofre) [reply sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: Done

File: sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java (hunk 3 above)
Comment (jbonofre) [reply sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java]: Agree. I'm figuring out a new to use the channel with the `ParquetWriter` to be use in `open()` method.

**Comment by Igosuki** [comment]
@eugene Parquet can be deserialized in several formats including Avro, I
don't think it's a good idea to tie it to Avro, instead users should be
able to use the Parquet read support implementations to map to values of
their choosing be it Avro or Protobuf messages.

**Comment by Igosuki** [comment]
Although it's likely most of the users today would use Generic Records as
the main Serde format of Parquet

**Comment by jkff** [comment]
@igosuki Thanks, that's a reasonable point, do you have a suggestion on how to incorporate this into the current API? Is it applicable on the write side too, I suppose?

I looked at the ReadSupport and WriteSupport classes. They seem quite tied to the Hadoop API and I'd prefer to not expose anything related to Hadoop from the Beam API surface - maybe we should just implement the main case of `readAvro()` for now, hiding the Hadoop details under the hood, and meanwhile think how to expose it more generally?

**Comment by Igosuki** [comment]
@jkff Likely something like :
```
myParquetIO.withReadSupport(ReadSupport<T>)
```
Where T has to implicitly respect the same extend clause as the one of the extending ReadSupport class so something like :
```
ParquetIo
.read("hdfs://myparquet")
.withReadSupport(ProtobufReadSupport.class, MyMessage.class)
...
.withProtobufCoder(ProtobufCoder.of(MyMessage))
```
Then it would yield here https://github.com/apache/beam/pull/1851/files#diff-03b8632d3bc8da69525461ba588f949bR330
```
static class ReadFn extends DoFn<FileIO.ReadableFile, GenericRecord> {
   @ProcessElement
   public void processElement(ProcessContext processContext) throws Exception {
     Path path = new Path(processContext.element().getMetadata().resourceId().toString());
     try (ParquetReader<T> reader = new ParquetReader(path, readSupport);
       T read;
       while ((read = reader.read()) != null) {
         processContext.output(read);
       }
     }
   }
 }
```

**Comment by jbonofre** [comment]
Agree, but it doesn't solve the reading/writing part. For this part, I'm working on using low level surface Parquet API.

**Comment by sabhyankar** [comment]
I was looking at this PR as I hope to use this in the near future :) Would it be possible to include full ValueProvider support so this can be templatized easily? I noticed that some of the fields (schema?) in the Builder does not provide a ValueProvider option.

**Comment by jbonofre** [comment]
Yes, but the blocking point now is not about the format/value provider, it's the core reader/writer based on FileIO. I sent an e-mail on Parquet mailing list, and I'm evaluating a workaround implementing our own Parquet reader/writer.

**Comment by Igosuki** [comment]
It would be awesome if the parquet reader could split work according to parquet file row group indexes. Is this what you have in mind @jbonofre ?

**Comment by jbonofre** [comment]
@Igosuki exactly, that's the purpose !

**Comment by Fokko** [comment]
What's the progress here? I'm willing to help if necessary. Cheers

**Comment by jbonofre** [comment]
@Fokko did you see my message on the dev mailing list about Parquet ? Basically, I asked Parquet guys if it would be possible to have a new release to smoothly use Beam `FileIO`.

**Comment by jbonofre** [comment]
For all watchers on this PR, we have basically three approaches:
1. If you take a look on the first commits in the PR, `ParquetIO` didn't leverage `FileIO` (because it didn't exist in the early stage). It was a complete IO with the "split" was limited (even if it worked fine for both read and write).
2. To be able to leverage `FileIO`, we need a way to use NIO Channel for Read. That's possible (via `InputStream`) with Parquet master as they expose the API (take a look on the mail I sent on the dev mailing list for details). This approach requires a new Parquet release.
3. Write our own Parquet reader/writer (low level, a bit like we did for Avro). It requires more work and I'm not so comfortable with this approach especially to leverage features and fixes provided by Parquet.

So, basically, I would love to hear your thoughts about `ParquetIO`: would you like a first simple `ParquetIO` in Beam even if the split is not optimal or if it's not based on `FileIO`, or are you OK for wait to have a `FileBased` one ?

By the way, I'm resuming my work on this one now that I'm back from travels.

**Comment by Igosuki** [comment]
I already tested this code in production, and another approach using Hadoop
input file IO and it works but not as well as delimiter based file IOs.
We should be able to create splits based on row groups.
I think what you need is something leveraging the latest features from S3A
(fadvise, head skip) and splits with hints for row groups, and obviously
including predicates and projections.
Right now for me, it's not satisfactory because with large parquet files
batch processes quickly become bottlenecked and parallelism is very
limited. It's frustrating that Spark is still better at this.

**Comment by jbonofre** [comment]
As a first approach, we can mimic what Spark does here.

**Comment by jbonofre** [comment]
@lgajowy by the way, does it mean that you already used the first "stage" of this PR ? I'm curious ;)

**Comment by Igosuki** [comment]
@jbonofre I did when you started it, once to create a Druid segment
rewriter and another to test performance of a simple windowed sum over a
large number of keys vs Spark.

My files go from small files (due to people having implemented constant
partitionning with different sized keys) to 400Mo parquet files.

I tested your code vs avro io vs the recent Hadoop input split over a few
TBs.

In the end, I had to fix coders which catch EOF because Flink throws EOF to
trigger spill on disk and reading parquet quickly overflows the heap...

The code works fine, but Hadoop input format io over S3A pefrozm better. I
think we need some examples to attain the same performance as Spark group
by + partition when there is a large cardinality over keys reading binary
formats.

**Comment by jbonofre** [comment]
@lgajowy thanks a lot for the details !!!! Do you mind to quickly describe (with a pipeline example) your "perfect target" ParquetIO ? I can implement that way as a first step. Thoughts ?

**Comment by dhalperi** [comment]
Reviewed 2 of 5 files at r1, 2 of 3 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/apache/beam/1851)*

**Comment by jbonofre** [comment]
I'm closing this PR as it has been included in the one from @lgajowy
